records/2022-07-20.security.prefix-api-token.md

```diff
@@ -0,0 +1,39 @@
+# TITLE
+
+Date: 2022-07-20
+
+## Status
+
+**DRAFT**
+
+## Context
+
+The mishandling of application secrets has lead to many security incidents in companies big and small. Given the inherent risk in how users store these secrets, it is important that we provide mechanisms to protect our users from doing so improperly.  Taking a cue from [Github](https://github.blog/2021-04-05-behind-githubs-new-authentication-token-formats/), [Slack](https://api.slack.com/authentication/token-types), and [Stripe](https://stripe.com/docs/api/authentication), we see that token prefixes are an effective technique to help us ensure our tokens don't end up in public repositories.
+
+By combining a known prefix with Github's [secret scanning program](https://docs.github.com/en/developers/overview/secret-scanning-partner-program), we can be alerted if one of our users mistakenly stores their API keys in the open and we can take action.
+
+## Options
+
+- Add a known prefix to our API keys in order to be able to easily identify them
+- Register with Github's secret scanning program and create an endpoint for them to report issues back to us
+- Give users a grace period to generate new keys
+  - After grace period, enforce a rule that ensures all API keys contain our prefix
```

Comment:
Do we want to add key expiration at the same time? ie so we can tell users in one message:
- you have to change your keys now
- from now on you have to change them every year (or whatever)

This would make the coding changes more complicated, but the user interactions easier. 🤷‍♂️

Reply:
Good idea and a good question.  @jimleroyer do you want key expiration to be part of this?

Reply:
It's a good idea but outside the scope of current work, i.e. make Notify API keys scanning possible within public repositories.